crates/adapters/src/controller.rs

```diff
@@ -1432,97 +1422,148 @@ struct CircuitThread {
     last_commit_progress_update: Instant,
 }
 
-/// Compare two maps for equality using a custom comparator for values.
-fn maps_equal_with<F, K, V>(a: &BTreeMap<K, V>, b: &BTreeMap<K, V>, mut cmp: F) -> bool
-where
-    K: Ord,
-    F: FnMut(&V, &V) -> bool,
-{
-    if a.len() != b.len() {
-        return false;
+/// Detect the following situation:
+///
+/// - A tables has _NOT_ been modified since the last checkpoint. This means that
+///   potentially some of the view that depend on it have not been modified either
+///   and will not be recomputed as part of the backfill. (NOTE: we don't actually
+///   check whether such views exist)
+/// - The table is marked as needing backfill in `bootstrap_info`. This is possible
+///   if at least one view that depends on this tables _HAS_ changed, however the
+///   table is not materialized, and so the only way to reconstruct that view is to
+///   ingest the contents of the table again.
+///
+/// In this situation we need to reject the pipeline bootstrap, as we don't currently
+/// have a way to replay input connectors UPTO the checkpointed state (we can fast
+/// forward them to this state, but not replay upto it) and then activate the rest of
+/// the circuit.
+///
+/// Is there another way? Yes. We could traverse the graph forward from such a table
+/// and mark all views derived from it for backfill, regardless of whether they have
+/// changed or not. This would ensure that the table is re-ingested, and all views
+/// derived from it are recomputed. We may want to add that later, but it seems like
+/// a good idea to push users towards materializing all tables. In fact I'm wondering
```

Comment:
this will not work for huge historical tables. it's nice to have a database, but some inputs are streaming by nature.